ratelimit/src/lib.rs

```diff
@@ -70,39 +79,52 @@ impl Ratelimiter {
     /// use ratelimit::*;
     ///
     /// // ratelimit to 1/s with no burst capacity
-    /// let ratelimiter = Ratelimiter::new(1, 1, 1);
+    /// let ratelimiter = Ratelimiter::new(1, 1, 1).unwrap();
     ///
     /// // ratelimit to 100/s with bursts up to 10
-    /// let ratelimiter = Ratelimiter::new(10, 1, 100);
+    /// let ratelimiter = Ratelimiter::new(10, 1, 100).unwrap();
     /// ```
-    pub fn new(capacity: u64, quantum: u64, rate: u64) -> Self {
-        let tick = SECOND / (rate / quantum);
-        Self {
+    pub fn new(capacity: u64, quantum: u64, rate: u64) -> Result<Self, Error> {
+        if rate == 0 {
+            return Err(Error::InvalidRate);
+        }
```

Comment:
There are still a bunch of cases where the ratelimiter parameters are invalid. I think that rate >= 1 B/s would not work in this implementation.

Curious what your use-case is and why you're looking to use this crate as opposed to some of the other implementations out there. Will be helpful as I consider the right changes to make in this crate.

Reply:
@brayniac I found this crate because it was one of the first result on crates.io and looked mature (a couple published versions - few issues). It has lots of downloads so I am obviously not the only one. Other options (like governor) may be harder to find just because of naming ...

I still like the fact that your implementation is smaller, simpler and easier to understand. This is why I took the time to report the issues I found while looking at it. Right now we have switched to another implementation but may switch back once the bug are fixed ;)

My use-case is rate limiting incoming messages (logs) before writing them to disk. For our use-case a big requirement is that the bucket must start filled (not empty) to allow burst from the beginning. I started working on a PR when I noticed the other issues.

Reply:
@sarfata - makes sense! The crate does have a long history and I use it mainly for rate-limiting request generation or operations in load generation / benchmarking cases.

I do appreciate you taking the time to report those issues, they're not things I've run into with my own usage.

One thing I'm considering that may make this even simpler is removing the `quantum` parameter and determining it automatically based on the rate. There's a lot of interplay between it and the rate and capacity that makes things like dynamically changing the rate potentially nasty. We can eliminate a lot of bad configuration possibilities if we set quantum dynamically for rates > 1B/s and just bump up the capacity to be the max of capacity and the quantum needed to achieve the desired rate. I think expressing the rate independent of quantum is probably easier to manage anyway.

Adding the option to start the bucket full instead of empty should be pretty easy. We can add a builder and make that configurable.

Reply:
> One thing I'm considering that may make this even simpler is removing the quantum parameter and determining it automatically based on the rate. There's a lot of interplay between it and the rate and capacity that makes things like dynamically changing the rate potentially nasty. We can eliminate a lot of bad configuration possibilities if we set quantum dynamically for rates > 1B/s and just bump up the capacity to be the max of capacity and the quantum needed to achieve the desired rate. I think expressing the rate independent of quantum is probably easier to manage anyway.

👍 I think it will also make the options easier to understand.